### PR TITLE
drivers: clock_control: add Ambiq clock control driver

### PR DESCRIPTION
**Pull request by aaronyegx**
This PR basically implements clock control driver for Ambiq Apollo series soc.
The clock control driver is mainly used to enable/disable the XTAL32MHz and XTAL32kHz and output it via specific IO to the dedicated modules, for example, the BLE controller inside of Apollo4 Blue Plus soc.
It is enabled in Apollo4 Blue Plus.

Verified the below Zephyr standard APIs worked as expected in apollo4p_blue_kxr_evb:
- clock_control_configure()
- clock_control_get_rate()
- clock_control_on()
- clock_control_off()

The configured clock IOs could output the accurate 32MHz and 32kHz clock and made the connected module (e.g., the BLE controller) work properly.

This PR should be merged after the HAL PR https://github.com/zephyrproject-rtos/hal_ambiq/pull/8 merged.

### CODE REVIEW

**Comment by aaronyegx** [comment]
Hi @nordic-krch @MaureenHelm @carlescufi @galak @nashif, could you help take a look at this PR? Thank you very much.

**Comment by aaronyegx** [comment]
Hi @MaureenHelm @fkokosinski @tgorochowik, if possible, could you take a look at this pull request? Thanks so much.

**Comment by aaronyegx** [comment]
Hi @nordic-krch , could you help talk a look and let's know what I can do to move forward this PR? Thanks.

File: include/zephyr/drivers/clock_control/clock_control_ambiq.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Copyright (c) 2023 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_AMBIQ_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_AMBIQ_H_
+
+#include <zephyr/device.h>
+#include <zephyr/sys/onoff.h>
```
Comment (nordic-krch) [inline include/zephyr/drivers/clock_control/clock_control_ambiq.h:11]: all includes are redundant.

**Review by nordic-krch — DISMISSED** [review]
Sorry for the delay. It looks ok with one minor comment.

File: include/zephyr/drivers/clock_control/clock_control_ambiq.h (hunk 1 above)
Comment (aaronyegx) [reply include/zephyr/drivers/clock_control/clock_control_ambiq.h]: @nordic-krch, thanks for your review. I clean up the redundant codes.

File: dts/arm/ambiq/ambiq_apollo4p_blue.dtsi [hunk 2]
```diff
@@ -11,6 +11,16 @@
 			clock-frequency = <24000000>;
 			#clock-cells = <0>;
 		};
+		xo32m: xo32m {
+			compatible = "ambiq,clkctrl";
+			clock-frequency = <32000000>;
```
Comment (MaureenHelm) [inline dts/arm/ambiq/ambiq_apollo4p_blue.dtsi:16]:
```suggestion
			clock-frequency = <DT_FREQ_M(32)>;
```

File: dts/arm/ambiq/ambiq_apollo4p_blue.dtsi [hunk 3]
```diff
@@ -11,6 +11,16 @@
...
+			#clock-cells = <1>;
+		};
+		xo32k: xo32k {
+			compatible = "ambiq,clkctrl";
+			clock-frequency = <32000>;
```
Comment (MaureenHelm) [inline dts/arm/ambiq/ambiq_apollo4p_blue.dtsi:21]:
```suggestion
			clock-frequency = <DT_FREQ_K(32)>;
````

**Review by MaureenHelm — DISMISSED** [review]

**Review by nordic-krch — DISMISSED** [review]

**Comment by aaronyegx** [comment]
Thanks @fkokosinski for merging the corresponding HAL PR.
Hi @nordic-krch @MaureenHelm , thanks for your review and approving. I addressed your comments and updated the west.yaml with latest HAL revision. Could you take another look at this, please?

**Review by nordic-krch — DISMISSED** [review]

**Comment by aaronyegx** [comment]
I had to rebase the branch to solve the conflict. @MaureenHelm @nordic-krch Looking forward to your revisiting, sorry for inconvenience.

**Review by nordic-krch — DISMISSED** [review]

**Comment by aaronyegx** [comment]
@nordic-krch Sorry I had to rebase again to solve new conflicts for this pending PR.

Hi @fkokosinski , could you take a look as well to help move forward this PR? Thanks very much.

**Comment by zchen-ambiq** [comment]
Hi, this is Zhongyi Chen from Ambiq. We could completely understand how busy you are with so many PRs to review. It'd greatly appreciated if you can help to review in time especially for those PRs sitting there a long time and their author addressing quickly to your comment. @nordic-krch @fabiobaltieri  Thank you in advance.

**Comment by aaronyegx** [comment]
Hi @MaureenHelm , could you help revisit, please? Thank you.

**Review by nordic-krch — DISMISSED** [review]

File: drivers/clock_control/clock_control_ambiq.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	if (clock_name >= CLOCK_CONTROL_AMBIQ_TYPE_MAX) {
+		return -EINVAL;
+	}
+
+	switch (clock_name) {
+	case CLOCK_CONTROL_AMBIQ_TYPE_HFXTAL_BLE:
+		arg.ui32_arg_hfxtal_user_mask = 1 << AM_HAL_HFXTAL_BLE_CONTROLLER_EN;
+		arg.b_arg_enable_HfXtalClockout = true;
+		ret = am_hal_mcuctrl_control(AM_HAL_MCUCTRL_CONTROL_EXTCLK32M_KICK_START,
+					     (void *)&arg);
```
[the remark this answers is not in the document]
Comment (aaronyegx) [reply drivers/clock_control/clock_control_ambiq.c:45]: @fabiobaltieri Yes we need this, otherwise it will cause -Wdiscarded-qualifiers compiling warning.

**Comment by aaronyegx** [comment]
Hi @fabiobaltieri thanks for your review, I addressed your comments.

File: drivers/clock_control/clock_control_ambiq.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/clock_control/clock_control_ambiq.c:45]: Uh, can you paste the full error and how you are testing it? tried to patch it and reproduce it locally but I'm not getting any. More of a curiosity than anything else, trying to understand what's behind it.

[the remark this answers is not in the document]
Comment (aaronyegx) [reply drivers/clock_control/clock_control_ambiq.c]: Hi @fabiobaltieri , thanks for your following. I got the compiling warning in another project with GCC compiler, so I kept the code consistent. I tried again in Zephyr clang environment locally (using Bluetooth samples with HCI Ambiq driver) and did not get warning. I'm OK to drop the casting to simplify the code here. Thanks.

/usr/bin/make project
../../../../../../ambiq_ble/apps/dtm_module/dtm_main.c:862:70: warning: passing argument 2 of 'am_hal_mcuctrl_control' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  862 |     am_hal_mcuctrl_control(AM_HAL_MCUCTRL_CONTROL_EXTCLK32M_DISABLE, &g_amHalMcuctrlArgBLEDefault);
      |                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../../../../mcu/apollo4p/am_mcu_apollo.h:156,
                 from ../../../../../../ambiq_ble/apps/dtm_module/dtm_main.c:23:
../../../../../../mcu/apollo4p/hal/mcu/am_hal_mcuctrl.h:452:46: note: expected 'void *' but argument is of type 'const am_hal_mcuctrl_control_arg_t *'
  452 |                                        void *pArgs);
      |

[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/clock_control/clock_control_ambiq.c]: Brilliant, yeah so the `discards 'const' qualifier` is actually a good warning to have that you'd suppress with that casting, so I think it's good to go without it. Thanks for checking it.

**Review by fabiobaltieri — APPROVED** [review]

**Comment by aaronyegx** [comment]
Hi @nordic-krch , could you help revisit again? Sorry to always dismiss your approving because of comment addressing and conflict resolving.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**